primefaces/src/main/java/org/primefaces/component/message/MessageRenderer.java

```diff
@@ -46,20 +49,30 @@ public void encodeEnd(FacesContext context, UIComponent component) throws IOExce
         Message uiMessage = (Message) component;
 
         UIComponent target = SearchExpressionUtils.contextlessResolveComponent(context, uiMessage, uiMessage.getFor());
-        String targetClientId = target.getClientId(context);
+        final MessageState state = new MessageState();
+        state.setTarget(target);
 
-        encodeMarkup(context, uiMessage, targetClientId);
-        encodeScript(context, uiMessage, target);
+        if (!hasDisplayableMessage(uiMessage, state.getTarget(), context) && CompositeUtils.isComposite(target)) {
+            ContextCallback callback = (FacesContext fc, UIComponent comp) -> {
+                if (hasDisplayableMessage(uiMessage, comp, fc)) {
+                    state.setTarget(comp);
```

Comment:
in theory you should extract all required infos from the component here (targetClientId and inputClientId) as e.g. EL with #{cc} is not resolveable outside this callback anymore